### PR TITLE
Enable mypy for ngclient

### LINKED ISSUE #1409: ngclient: type annotations review and mypy integration
ngclient type annotations are not complete: once mypy is in use (#1395), we should
* enable mypy for ngclient
* fix any annotation issues

### PR DESCRIPTION
**Pull request by sechkova**
Fixes #1409 
**Description of the changes being introduced by the pull request**:

Enable static type checking for ngclient.
- add all missing type annotations
- fix errors (partially for updater.py and trusted_metadata_set.py
- add urllib3 to ignored includes
- download suggested stub packages when running mypy

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by jku** [comment]
I think there's a few cases here:
* TrustedMetadataSet.root probably does not actually need to be Optional: Unlike other properties it's guaranteed to be a Metadata after constructor, right?
* mypy gets confused about target_base_url: I'm sure it's possible to simplify the code so mypy understands that target_base_url cannot be None
* cases where we know a e.g. self._trusted_set.snapshot is Metadata but code does not: I think this is a great place for `assert(self._trusted_set.snapshot is not None)` 
* many cases of Metadata.Signed typing problem...
  * I believe this would get fixed with #1457 and Generics
  * alternatively we could just use calls like `assert(isinstance(self._trusted_set.snapshot.signed, Snapshot))`. Ugly but works and there are surprisingly few places where this is needed (like 10-12 or so?)
  * even more alternatively we could add properties to TrustedMetadataSet that returned Root, Snapshot, Timestamp and Targets (but that makes a lot of the code a bit ugly as you usually need both the Metadata and the Signed subclass)...

First two-three can be fixed now. The last one we can discuss... In any case this is useful work as it shows how much (or little) the lack of signed typing affects us...

The download related the changes should maybe only be done after the cleanup is done in #1448 (also the SlowRetrievalError change looks a bit fishy -- I would expect it's used in the legacy code as well?)

**Comment by joshuagl** [comment]
> * mypy gets confused about target_base_url: I'm sure it's possible to simplify the code so mypy understands that target_base_url cannot be None

It's debatable whether this simplifies things for humans reading the code (edit: though, apart from the variable name ugliness and possibility for confusion of `tgt_base_url`, I think it probably is simpler) but I was able to make mypy happy with the following:

```Python
        # locally scoped variable that is definitely a string, even if empty
        tgt_base_url: str = self._target_base_url or ""
        if target_base_url is not None:
            tgt_base_url = _ensure_trailing_slash(target_base_url)

        # move the check for a base URL being set to after the locally scoped
        # variable assignment logic is bottomed out.
        if not tgt_base_url:
            raise ValueError(
                "target_base_url must be set in either download_target() or "
                "constructor"
            )
```

see the [full diff here](https://github.com/joshuagl/tuf/commit/2aaec55ab303bf9b753c863df835d9be50c4a2bd).

**Comment by jku** [comment]
this should also work -- the issue for mypy is two branches that seem unrelated (even if they aren't) so making the branches actually same should make the intent clear to computer...
```
        if target_base_url is None:
            if self._target_base_url is None:
                raise ValueError(
                    "target_base_url must be set in either download_target() or "
                    "constructor"
                )
            target_base_url = self._target_base_url
        else:
            target_base_url = _ensure_trailing_slash(target_base_url)
```

**Comment by jku** [comment]
> The download related the changes should maybe only be done after the cleanup is done in #1448 (also the SlowRetrievalError change looks a bit fishy -- I would expect it's used in the legacy code as well?)

#1448 is now merged

**Comment by sechkova** [comment]
Rebased on latest changes in develop and fixed all issues but the Metadata.Signed typing ones.

**Comment by sechkova** [comment]
Actually `bandit` complains slightly about asserts:
```
>> Issue: [B101:assert_used] Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.
   Severity: Low   Confidence: High
   Location: tuf/ngclient/updater.py:297
   More Info: https://bandit.readthedocs.io/en/latest/plugins/b101_assert_used.html
296	
297	            assert self._trusted_set.snapshot is not None
298	            metainfo = self._trusted_set.snapshot.signed.meta[f"{role}.json"]
```

**Comment by joshuagl** [comment]
> Actually `bandit` complains slightly about asserts

Right. We could reasonably silence those warnings, as we are only suggesting use of asserts to help mypy determine the expected type. But I think this further reinforces the 'code smell' of using asserts to solve this problem.

**Comment by sechkova** [comment]
> > Actually `bandit` complains slightly about asserts
> 
> Right. We could reasonably silence those warnings, as we are only suggesting use of asserts to help mypy determine the expected type. But I think this further reinforces the 'code smell' of using asserts to solve this problem.

In this case we use assert as the [recommended way](https://mypy.readthedocs.io/en/stable/common_issues.html#unexpected-errors-about-none-and-or-optional-types) to deal with `None` errors. So we either silence bandit or tweak the TrustedMetadataSet public properties to not return optional values.

**Comment by sechkova** [comment]
I used the part of #1457 that makes Metadata 'Generic' to resolve the type checking issues related to the 'signed' type.
Now there are no more mypy errors left and the PR should be ready for review.

File: tuf/api/metadata.py
Comment (joshuagl) [inline tuf/api/metadata.py:756]: `BinaryIO` appears to be a subclass of `IO[bytes]` with the addition of interfaces to support context managers (`with`). We are not actually using the passed `data` object as a context manager, so this looks fine.

File: tuf/ngclient/updater.py
Comment (joshuagl) [inline tuf/ngclient/updater.py:115]: Nice 👍

File: tuf/ngclient/_internal/trusted_metadata_set.py
Comment (joshuagl) [inline tuf/ngclient/_internal/trusted_metadata_set.py]: Nice catch! thanks to mypy?

File: tuf/ngclient/updater.py
Comment (joshuagl) [inline tuf/ngclient/updater.py]: I wonder if there is a better return type we could be using here? Rather than a dict which will only ever contain two key-value pairs where the values have different types?

I am not suggesting we change something here in this PR, just thinking out loud.

File: tuf/ngclient/_internal/trusted_metadata_set.py
Comment (joshuagl) [inline tuf/ngclient/_internal/trusted_metadata_set.py]: If the root property is guaranteed to be set after init, do we need this check?

File: tuf/ngclient/_internal/trusted_metadata_set.py
Comment (joshuagl) [inline tuf/ngclient/_internal/trusted_metadata_set.py]: These type hints for the IDE are starting to look quite cumbersome. Can the IDE not recognise the type hint on the right-hand side of the assignment operator? Do we really need them?😕

**Review by joshuagl — COMMENTED** [review]
Thanks Teodora. This is looking good. I have a minor comment on the change to the root property, and a slight pushback on the type hinting for IDE's that I would appreciate your thoughts on before we merge.

File: tuf/ngclient/_internal/trusted_metadata_set.py
Comment (sechkova) [reply tuf/ngclient/_internal/trusted_metadata_set.py]: Yes, mypy gets the credit.

File: tuf/ngclient/_internal/trusted_metadata_set.py
Comment (sechkova) [reply tuf/ngclient/_internal/trusted_metadata_set.py]: We need it because we call `update_root` during init too ... to validate the initial root.
Do you think we need to clarify it somehow?

File: tuf/ngclient/updater.py
Comment (sechkova) [reply tuf/ngclient/updater.py]: Yeah, I think there is a common "willingness" and a bunch of issues related to improving this part of the API.
The suggestion here seems to be the closest to yours: https://github.com/theupdateframework/tuf/issues/1411

File: tuf/ngclient/_internal/trusted_metadata_set.py
Comment (sechkova) [reply tuf/ngclient/_internal/trusted_metadata_set.py]: Now that you made me have another look :roll_eyes: .... it turned out I wasn't able to annotate the return type of the `from_*` methods with the generic type 'T' because `from_bytes` was a static method. So I amended the last commit and I think this one now looks better and does the job: 6754470

**Comment by sechkova** [comment]
Seems like the CI won't be triggered until I rebase and resolve the conflicts.

File: tuf/ngclient/_internal/trusted_metadata_set.py
Comment (sechkova) [reply tuf/ngclient/_internal/trusted_metadata_set.py]: Since you expressed some worries about inconsistently using the 'root' property, the other solution I came up with is this  https://github.com/theupdateframework/tuf/pull/1489/commits/8c172af244c968b221bcbd5398a7a07a6c897408

**Comment by jku** [comment]
> * add urllib3 to ignored includes

If you researched this can you document why? looking in urllib3 git the annotations seem to be there but is that just for some future 2.0 release?

> * download suggested stub packages when running mypy

Could we add the required stub packages to test requirements instead?

File: tuf/ngclient/_internal/trusted_metadata_set.py
Comment (jku) [inline tuf/ngclient/_internal/trusted_metadata_set.py]:
```suggestion
        self._trusted_set: Dict[str, Metadata] = {}
```
Since we're fixing my broken annotations, let's do it like everywhere else :grimacing:

File: tuf/ngclient/_internal/requests_fetcher.py
Comment (jku) [inline tuf/ngclient/_internal/requests_fetcher.py]: Reasons for this are good but now the documented exceptions (here and in Fetcher API) are incorrect... Hmm. Maybe you could make SlowRetrievals speed argument optional instead?

**Review by jku — COMMENTED** [review]
I think it looks good (but obviously the last commits depend on a commit from generics PR #1457 so I'm not approving). 

This PR isn't using the whole Generics branch though: if it did then it could just remove all of the checks of type `if new_root.signed.type != "root"` since the constructor would do that already

The last commit is IMO not great since it has so much code in the try-block: it's not clear that the KeyError should only happen in `self.root`... I think the original code is better. Maybe the real alternative would be to not call `update_root()` from init but instead do the required steps in init itself -- that seems ok to me as well as I think there aren't that many steps here.

File: tuf/ngclient/_internal/requests_fetcher.py
Comment (jku) [reply tuf/ngclient/_internal/requests_fetcher.py]: or figure out another solution -- I didn't spend a lot of time on that suggestion

**Comment by sechkova** [comment]
> I think it looks good (but obviously the last commits depend on a commit from generics PR #1457 so I'm not approving).

Hmm ... I wouldn't say directly depending, I didn't cherry-pick the commit from the other PR but created a new one and added you as a co-author.
 
> This PR isn't using the whole Generics branch though: if it did then it could just remove all of the checks of type `if new_root.signed.type != "root"` since the constructor would do that already

This was done intentionally but I guess I should've tried to explain my intentions somewhere in written form. 

The "Generics" branch tries to achieve two things simultaneously: correct type annotations of Signed and a runtime check ensuring  content and type are really what we expect. Since there seems to be an agreement already on the use of Generics and an ongoing discussion around the runtime check part, I decided to include only the first part here, the one directly related to type annotations and mypy. This way we can leave the rest of the discussion for #1457 which will also make it simpler because it will try to solve only one problem. I hope you agree :)

**Comment by sechkova** [comment]
> The last commit is IMO not great since it has so much code in the try-block: it's not clear that the KeyError should only happen in `self.root`... I think the original code is better. Maybe the real alternative would be to not call `update_root()` from init but instead do the required steps in init itself -- that seems ok to me as well as I think there aren't that many steps here.

I tried this alternative since @joshuagl expressed some concerns about not using the 'root' property consistently inside the class. The three options, none of them perfect, would be:
- don't call `update_root` from `init`, repeat some of the code (not than much repetition so could be acceptable)
- don't use the property but "get" safely from `self._trusted_set`  c205ac0 (concerns about not using consistently the root property)
- use the root property, enclose it in a try bock and catch the KeyError 8c172af (well you already described the disadvantages)

There is a forth one where we keep root optional as the rest of the properties and add asserts in the code but I tried and this means asserts at 4-5 places which seems to rule out this option.

**Comment by jku** [comment]
> > I think it looks good (but obviously the last commits depend on a commit from generics PR #1457 so I'm not approving).
> 
> Hmm ... I wouldn't say directly depending, I didn't cherry-pick the commit from the other PR but created a new one and added you as a co-author.
> 
> > This PR isn't using the whole Generics branch though: if it did then it could just remove all of the checks of type `if new_root.signed.type != "root"` since the constructor would do that already
> 
> This was done intentionally but I guess I should've tried to explain my intentions somewhere in written form.

Yeah I think the generics change is major enough that we can't just hide it in a PR titled "enable mypy for ngclient" :) We can definitely split the PR if that makes it easier to swallow though.

**Comment by joshuagl** [comment]
> Yeah I think the generics change is major enough that we can't just hide it in a PR titled "enable mypy for ngclient" :) We can definitely split the PR if that makes it easier to swallow though.

Ack, let's do it.

**Comment by sechkova** [comment]
> > * add urllib3 to ignored includes
> 
> If you researched this can you document why? looking in urllib3 git the annotations seem to be there but is that just for some future 2.0 release?

Yes, the latest release is 1.26 which does not include the type annotations. I updated the commit description.

> > * download suggested stub packages when running mypy

> Could we add the required stub packages to test requirements instead?

Done:  d306370

File: tuf/ngclient/_internal/requests_fetcher.py
Comment (sechkova) [reply tuf/ngclient/_internal/requests_fetcher.py]: Yep, optional seems to fit well: 7cbfd0e

**Comment by jku** [comment]
Ok, I've made the #1457 PR into a "minimal generics" PR -- this PR could be based off of that one I believe.

**Comment by sechkova** [comment]
Rebased on #1457 and I believe I have addressed all comments so waiting for another review.

File: tuf/ngclient/updater.py
Comment (jku) [inline tuf/ngclient/updater.py:344]: this is unfortunate... I don't have any better ideas but it looks so strange in this code.

This seems to only be needed in Updater for getting meta-info -- which is also a sort of duplicate piece of code here and in `_load_targets()`... so there's probably some sort of refactoring possibility here but let's not do that in this PR.

File: tuf/api/metadata.py
Comment (jku) [inline tuf/api/metadata.py:101]: I don't object but I'm curious did this actually affect something? I expect from_dict() to be only called from Deserializer which does not know about specific signed types.

**Review by jku — APPROVED** [review]
Looks good, updater.py in VSCode is much nicer now. I think this is good to go... except this will definitely clash with #1514 which I was about to merge...

File: tuf/ngclient/_internal/trusted_metadata_set.py
Comment (jku) [reply tuf/ngclient/_internal/trusted_metadata_set.py]: this looks good to me at least

File: tuf/api/metadata.py
Comment (sechkova) [reply tuf/api/metadata.py:101]: No, I did it just for consistency with the others.

**Comment by sechkova** [comment]
Rebased on latest changes in develop.

**Pull request merged**
